Make FunctionalTensor subclass to be more like functorch (interaction with ZeroTensor + Conjugate key)

bdhirsh opened the pull request:
I added some tests for Conj, Neg and ZeroTensor for both python and C++ functionalization. This also fixes a nasty segfult when running a functorch `jacfwd` test with `torch.compile`, once AOTAutograd is using `FunctionalTensor`.

Changes:

(1) I use Jeffrey's `make_wrapper_subclass(extra_dispatch_keys)` kwarg to plumb extra dispatch keys ontoto the wrapper, mirroring what C++ functionalization does (C++ functionalization will mirror all dispatch keys from the inner tensor to the wrapper, except for python and functorch keys).

(2) FunctionalTensorMode will decompose CompositeImplicitAutograd ops, since (for example) ZeroTensor kernels can send ops like `.to()` directly to the Python key. We'll need a way to toggle this later for pre-dispatch functionalization

(3) Bound `_ForceDispatchKeyGuard` and BatchedTensorImpl's dispatch keyset to python

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #105236
* #104974
* #104483
* #106406
* __->__ #109023
* #109632
* #109662
* #108654

bdhirsh committed:
Fix wrapper subclass interaction with ZeroTensor + Conjugate key

[ghstack-poisoned]

bdhirsh committed:
Update on "Fix wrapper subclass interaction with ZeroTensor + Conjugate key"




[ghstack-poisoned]

bdhirsh commented:
After talking to Richard, I'm going try to find the API that let's you reset TLS to whatever the PythonTLSSnapshot key stashed, and use that inside of functional_tensor.py

ezyang commented:
You want `enable_reentrant_dispatch`

ezyang commented:
Brian has already tried this, see https://github.com/pytorch/pytorch/issues/108983#issuecomment-1714706014

bdhirsh committed:
Update on "Fix wrapper subclass interaction with ZeroTensor + Conjugate key"

This is a potential fix for https://github.com/pytorch/pytorch/issues/108983. It fixes an ugly segfault that I got when running a functorch `jacfwd()` test with `torch.compile`, once I started using `FunctionalTensor` in AOTAutograd.

Notable changes:

(1) register a fallthrough for ZeroTensor to `prim::clone`, which is what we do for `aten::clone`. We need this now if we expect the `clone()` decomp that we use to work with ZeroTensor's.

(2) `maybe_add_maybe_sub` is a kernel used by ZeroTensor that does some meta-tensor compute to figure out the size of the ZeroTensor to return. This broke in my tests because we end up dispatching `aten::to()` directly to my (FunctionalTensor) subclass. This is not ideal, because `FunctionalTensor` uses `return_and_correct_aliasing()` to properly handling storage aliasing, and to do that it reads aliasing information off of the schema of each op it sees. `to()` is a "maybe aliasing" op, and so we'd like to decompose it before getting to `_torch_dispatch_`

(3) I removed `ZeroTensor` (and conj and neg) from the exclude set in the python fallback kernel. This didn't feel too unreasonable... but I'm wondering if anyone has an opinion about it. An alternative would be to not do this, and instead to have `FunctionalTensor` manually remove these keys from the exclude set upon redispatch (aka put the onus on the tensor subclass to fix up TLS, as described in https://github.com/pytorch/pytorch/issues/108983).




[ghstack-poisoned]

bdhirsh committed:
Update on "Fix wrapper subclass interaction with ZeroTensor + Conjugate key"

This is a potential fix for https://github.com/pytorch/pytorch/issues/108983. It fixes an ugly segfault that I got when running a functorch `jacfwd()` test with `torch.compile`, once I started using `FunctionalTensor` in AOTAutograd.

Notable changes:

(1) register a fallthrough for ZeroTensor to `prim::clone`, which is what we do for `aten::clone`. We need this now if we expect the `clone()` decomp that we use to work with ZeroTensor's.

(2) `maybe_add_maybe_sub` is a kernel used by ZeroTensor that does some meta-tensor compute to figure out the size of the ZeroTensor to return. This broke in my tests because we end up dispatching `aten::to()` directly to my (FunctionalTensor) subclass. This is not ideal, because `FunctionalTensor` uses `return_and_correct_aliasing()` to properly handling storage aliasing, and to do that it reads aliasing information off of the schema of each op it sees. `to()` is a "maybe aliasing" op, and so we'd like to decompose it before getting to `_torch_dispatch_`

(3) I removed `ZeroTensor` (and conj and neg) from the exclude set in the python fallback kernel. This didn't feel too unreasonable... but I'm wondering if anyone has an opinion about it. An alternative would be to not do this, and instead to have `FunctionalTensor` manually remove these keys from the exclude set upon redispatch (aka put the onus on the tensor subclass to fix up TLS, as described in https://github.com/pytorch/pytorch/issues/108983).




[ghstack-poisoned]

bdhirsh commented:
I updated this PR to do a more targeted fix for FunctionalTensor. I think that if we want e.g. `TwoTensor`/`DTensor` and other subclasses to work well with both the ZeroTensor and Conjugate keys (e.g. users that use complex dtypes and fw-AD), we'll need to think more about how this should work. But since my main goal is just to turn on python functionalization in aot autograd, I'm ok with some more one-off fixes for functionalization.

changes in the PR are now:

(1) to support ZeroTensor, I manually remove the ZeroTensor key from the exclude TLS in `FunctionalTensor.__torch_dispatch__`

(2) to support properly handling `is_conj()` and `is_neg()` calls on my `FunctionalTensor`, I make sure to mirror the dispatch key set from my inner tensor onto my `FunctionalTensor` wrapper, specifically if they have the `Conjugate` or `Negative` dispatch keys. It was actually pretty easy to do this thanks to @soulitzer 's new `_extra_dispatch_keys` kwarg.

(3) added some testing in `test_functionalization.py`

bdhirsh committed:
Update on "Fix wrapper subclass interaction with ZeroTensor + Conjugate key"

This is a potential fix for https://github.com/pytorch/pytorch/issues/108983. It fixes an ugly segfault that I got when running a functorch `jacfwd()` test with `torch.compile`, once I started using `FunctionalTensor` in AOTAutograd.

Notable changes:

(1) register a fallthrough for ZeroTensor to `prim::clone`, which is what we do for `aten::clone`. We need this now if we expect the `clone()` decomp that we use to work with ZeroTensor's.

(2) `maybe_add_maybe_sub` is a kernel used by ZeroTensor that does some meta-tensor compute to figure out the size of the ZeroTensor to return. This broke in my tests because we end up dispatching `aten::to()` directly to my (FunctionalTensor) subclass. This is not ideal, because `FunctionalTensor` uses `return_and_correct_aliasing()` to properly handling storage aliasing, and to do that it reads aliasing information off of the schema of each op it sees. `to()` is a "maybe aliasing" op, and so we'd like to decompose it before getting to `_torch_dispatch_`

(3) I removed `ZeroTensor` (and conj and neg) from the exclude set in the python fallback kernel. This didn't feel too unreasonable... but I'm wondering if anyone has an opinion about it. An alternative would be to not do this, and instead to have `FunctionalTensor` manually remove these keys from the exclude set upon redispatch (aka put the onus on the tensor subclass to fix up TLS, as described in https://github.com/pytorch/pytorch/issues/108983).




[ghstack-poisoned]

bdhirsh committed:
Update on "Fix wrapper subclass interaction with ZeroTensor + Conjugate key"

This is a potential fix for https://github.com/pytorch/pytorch/issues/108983. It fixes an ugly segfault that I got when running a functorch `jacfwd()` test with `torch.compile`, once I started using `FunctionalTensor` in AOTAutograd.

Notable changes:

(1) register a fallthrough for ZeroTensor to `prim::clone`, which is what we do for `aten::clone`. We need this now if we expect the `clone()` decomp that we use to work with ZeroTensor's.

(2) `maybe_add_maybe_sub` is a kernel used by ZeroTensor that does some meta-tensor compute to figure out the size of the ZeroTensor to return. This broke in my tests because we end up dispatching `aten::to()` directly to my (FunctionalTensor) subclass. This is not ideal, because `FunctionalTensor` uses `return_and_correct_aliasing()` to properly handling storage aliasing, and to do that it reads aliasing information off of the schema of each op it sees. `to()` is a "maybe aliasing" op, and so we'd like to decompose it before getting to `_torch_dispatch_`

(3) I removed `ZeroTensor` (and conj and neg) from the exclude set in the python fallback kernel. This didn't feel too unreasonable... but I'm wondering if anyone has an opinion about it. An alternative would be to not do this, and instead to have `FunctionalTensor` manually remove these keys from the exclude set upon redispatch (aka put the onus on the tensor subclass to fix up TLS, as described in https://github.com/pytorch/pytorch/issues/108983).




[ghstack-poisoned]

bdhirsh committed:
Update on "Fix wrapper subclass interaction with ZeroTensor + Conjugate key"

This is a potential fix for https://github.com/pytorch/pytorch/issues/108983. It fixes an ugly segfault that I got when running a functorch `jacfwd()` test with `torch.compile`, once I started using `FunctionalTensor` in AOTAutograd.

Notable changes:

(1) register a fallthrough for ZeroTensor to `prim::clone`, which is what we do for `aten::clone`. We need this now if we expect the `clone()` decomp that we use to work with ZeroTensor's.

(2) `maybe_add_maybe_sub` is a kernel used by ZeroTensor that does some meta-tensor compute to figure out the size of the ZeroTensor to return. This broke in my tests because we end up dispatching `aten::to()` directly to my (FunctionalTensor) subclass. This is not ideal, because `FunctionalTensor` uses `return_and_correct_aliasing()` to properly handling storage aliasing, and to do that it reads aliasing information off of the schema of each op it sees. `to()` is a "maybe aliasing" op, and so we'd like to decompose it before getting to `_torch_dispatch_`

(3) I removed `ZeroTensor` (and conj and neg) from the exclude set in the python fallback kernel. This didn't feel too unreasonable... but I'm wondering if anyone has an opinion about it. An alternative would be to not do this, and instead to have `FunctionalTensor` manually remove these keys from the exclude set upon redispatch (aka put the onus on the tensor subclass to fix up TLS, as described in https://github.com/pytorch/pytorch/issues/108983).




[ghstack-poisoned]

bdhirsh committed:
Update on "Fix FunctionalTensor wrapper subclass interaction with ZeroTensor + Conjugate key"

This is a potential fix for https://github.com/pytorch/pytorch/issues/108983. It fixes an ugly segfault that I got when running a functorch `jacfwd()` test with `torch.compile`, once I started using `FunctionalTensor` in AOTAutograd.

Notable changes:

(1) register a fallthrough for ZeroTensor to `prim::clone`, which is what we do for `aten::clone`. We need this now if we expect the `clone()` decomp that we use to work with ZeroTensor's.

(2) `maybe_add_maybe_sub` is a kernel used by ZeroTensor that does some meta-tensor compute to figure out the size of the ZeroTensor to return. This broke in my tests because we end up dispatching `aten::to()` directly to my (FunctionalTensor) subclass. This is not ideal, because `FunctionalTensor` uses `return_and_correct_aliasing()` to properly handling storage aliasing, and to do that it reads aliasing information off of the schema of each op it sees. `to()` is a "maybe aliasing" op, and so we'd like to decompose it before getting to `_torch_dispatch_`

(3) I removed `ZeroTensor` (and conj and neg) from the exclude set in the python fallback kernel. This didn't feel too unreasonable... but I'm wondering if anyone has an opinion about it. An alternative would be to not do this, and instead to have `FunctionalTensor` manually remove these keys from the exclude set upon redispatch (aka put the onus on the tensor subclass to fix up TLS, as described in https://github.com/pytorch/pytorch/issues/108983).




[ghstack-poisoned]

ezyang commented:
deferring to @zou3519 entirely here

bdhirsh committed:
Update on "Make FunctionalTensor subclass to be more like functorch (interaction with ZeroTensor + Conjugate key)"

This is a potential fix for https://github.com/pytorch/pytorch/issues/108983. It fixes an ugly segfault that I got when running a functorch `jacfwd()` test with `torch.compile`, once I started using `FunctionalTensor` in AOTAutograd.

Notable changes:

(1) register a fallthrough for ZeroTensor to `prim::clone`, which is what we do for `aten::clone`. We need this now if we expect the `clone()` decomp that we use to work with ZeroTensor's.

(2) `maybe_add_maybe_sub` is a kernel used by ZeroTensor that does some meta-tensor compute to figure out the size of the ZeroTensor to return. This broke in my tests because we end up dispatching `aten::to()` directly to my (FunctionalTensor) subclass. This is not ideal, because `FunctionalTensor` uses `return_and_correct_aliasing()` to properly handling storage aliasing, and to do that it reads aliasing information off of the schema of each op it sees. `to()` is a "maybe aliasing" op, and so we'd like to decompose it before getting to `_torch_dispatch_`

(3) I removed `ZeroTensor` (and conj and neg) from the exclude set in the python fallback kernel. This didn't feel too unreasonable... but I'm wondering if anyone has an opinion about it. An alternative would be to not do this, and instead to have `FunctionalTensor` manually remove these keys from the exclude set upon redispatch (aka put the onus on the tensor subclass to fix up TLS, as described in https://github.com/pytorch/pytorch/issues/108983).




[ghstack-poisoned]

bdhirsh committed:
Update on "Make FunctionalTensor subclass to be more like functorch (interaction with ZeroTensor + Conjugate key)"

This is a potential fix for https://github.com/pytorch/pytorch/issues/108983. It fixes an ugly segfault that I got when running a functorch `jacfwd()` test with `torch.compile`, once I started using `FunctionalTensor` in AOTAutograd.

Notable changes:

(1) register a fallthrough for ZeroTensor to `prim::clone`, which is what we do for `aten::clone`. We need this now if we expect the `clone()` decomp that we use to work with ZeroTensor's.

(2) `maybe_add_maybe_sub` is a kernel used by ZeroTensor that does some meta-tensor compute to figure out the size of the ZeroTensor to return. This broke in my tests because we end up dispatching `aten::to()` directly to my (FunctionalTensor) subclass. This is not ideal, because `FunctionalTensor` uses `return_and_correct_aliasing()` to properly handling storage aliasing, and to do that it reads aliasing information off of the schema of each op it sees. `to()` is a "maybe aliasing" op, and so we'd like to decompose it before getting to `_torch_dispatch_`

(3) I removed `ZeroTensor` (and conj and neg) from the exclude set in the python fallback kernel. This didn't feel too unreasonable... but I'm wondering if anyone has an opinion about it. An alternative would be to not do this, and instead to have `FunctionalTensor` manually remove these keys from the exclude set upon redispatch (aka put the onus on the tensor subclass to fix up TLS, as described in https://github.com/pytorch/pytorch/issues/108983).




[ghstack-poisoned]

zou3519 commented on aten/src/ATen/ConjugateFallback.cpp:
```diff
@@ -18,6 +18,10 @@ TORCH_LIBRARY_IMPL(_, Conjugate, m) {
   m.fallback(torch::CppFunction::makeFromBoxedFunction<&conjugateFallback>());
 }
 
+TORCH_LIBRARY_IMPL(prims, Conjugate, m) {
+  m.impl("clone", torch::CppFunction::makeFallthrough());
+}
```
Add test please?

bdhirsh replied:
will do!

bdhirsh replied:
turns out that this wasn't even needed (I had to fix https://github.com/pytorch/pytorch/issues/103532 instead)

zou3519 commented on test/test_functionalization.py:
```diff
@@ -1542,6 +1542,35 @@ def forward(self, x_1):
     add = torch.ops.aten.add.Tensor(view_1, 1);  view_1 = None
     return add""")
 
+    def test_python_functionalization_zero_tensor(self):
```
We should probably also run these tests, but with C++ functionalization, to make sure the behavior is the same.

NB: Despite DispatchKey::ZeroTensor being ABOVE DispatchKey::Functionalize, what actually happens for C++ functionalization is that DispatchKey::Functionalize runs first! (Because we use Functionalize as a Mode, and it ends up wrapping the constructed EfficientZeroTensor) This is OK because one cannot do any in-place operations to EfficientZeroTensor and is what we want to happen with Python FunctionalTensor as well.

bdhirsh replied:
This is definitely fair, good call out. I'll add a test for both.

zou3519 commented on test/test_functionalization.py:
```diff
@@ -1542,6 +1542,35 @@ def forward(self, x_1):
+        self.assertEqual(out_ref, out_test)
+
+    def test_python_functionalization_conj(self):
+        def f(x):
+            y = x.conj()
+            return torch.view_as_real(y.resolve_conj())
```
Btw, C++ functionalization looks silently incorrect on .conj():
```
import torch
from torch._functorch.aot_autograd import from_fun, to_fun
import torch.utils._pytree as pytree
from functorch import make_fx

def cpp_functionalize(f):
    def inner(*args):
        fun_args = pytree.tree_map_only(torch.Tensor, to_fun, args)
        try:
            torch._enable_functionalization(reapply_views=False)
            fun_outs = f(*fun_args)
        finally:
            torch._disable_functionalization()
        outs = pytree.tree_map_only(torch.Tensor, from_fun, fun_outs)
        return outs
    return inner

def f(x):
    y = x.conj()
    return y

x = torch.randn(4)
gm = make_fx(cpp_functionalize(f))(x)
print(gm.code)
```
What does Python functionalization show for this case? The correct thing here seems like there should have been a call to .conj_physical.

bdhirsh replied:
Hmm... what about that output makes it silently wrong?

In my head, the relationship between `.conj() <-> .conj_physical()` is analogous to `.transpose() <-> transpose_copy()`. The first toggles some metadata on the tensor saying "this tensor is a view", while the second materializes that view - and only later, once we hit an op that needs the view to be materialized do we go and materialize it (e.g. a kernel that can't handle non-contiguous or conjugated views).

bdhirsh replied:
I guess one difference is that if we consider AOTDispatcher to be really ahead-of-time tracing out dispatcher functionalities, and `conj -> conj_physical` is implemented as a dispatcher functionality, then you're saying that we expect inductor to never have to worry about handling `conj`, and only have to worry about handling `conj_physical`?

zou3519 replied:
Sorry, forgot to paste the output:
```
def forward(self, arg0_1):
    return arg0_1
```
That's what gets traced. Since .conj is a view operation, I would expect that there be a call to conj_physical in the graph instead of no call at all

zou3519 reviewed (commented):
EfficientZeroTensor handling LGTM. It's a bit hard to visualize what is going on with conj/neg so I asked for some more detailed test cases

bdhirsh commented on aten/src/ATen/native/BinaryOps.cpp:
```diff
@@ -1054,8 +1054,17 @@ static Tensor maybe_add_maybe_sub(const Tensor& self, const Tensor& other, const
+  // We **need** to make sure that the CompositeImplicitAutograd to() kernel runs, so that
+  // if there are any subclasses, they always see the desguared to() call.
+  // We are in a ZeroTensor kernel below autograd, though, which means that `.to()`
+  // will hit the Python key before hitting the composite kernel.
```
So I hit this when I first wrote some basic manual tests for zero tensor (I *think* my test in `test_functionalization.py` actually exercises this, but let me double check and add a better test if not).

The problem is that this function is registered to the `ZeroTensor` dispatch key. The relevant dispatch keys are something like:
```
AutogradCPU (autograd has already run)
ZeroTensor <-- we are here
Python <-- we will dispatch here next
CPU
```

`aten::to()` has a `CompositeImplicitAutograd` registration. `CompositeImplicitAutograd` is an alias key that effectively registers the function to all autograd keys, and all "backend" keys (*not* including python, since python isn't a backend key).

So when we call `aten::to()` here, the next key we hit is the Python key, and we never get a chance to hit a key that runs the CompositeImplicitAutograd kernel.

bdhirsh replied:
Hmm, maybe a good question is why don't we just add the `Python` key to the `CompositeImplicitAutograd` alias keyset today? I'm not actually sure.

I'd be happy to try making that change. Although I'm a bit worried that it might be higher blast radius. If it's ok with you, I'm accumulating a list of followups that I want to revisit once I've unblocked DTensor compile support, and I can add this to the list. What do you think?

zou3519 replied:
Does C++ functionalization have the same behavior?

bdhirsh replied:
Hmm, we don't actually want to disable the python key (why? We still want FakeTensor to run as part of computing `meta_out` below, since we need it e.g. in case the output has symbolic shapes).

I think in theory we'd want something like a C++ variant of `disable_proxy_modes_tracing()` (to just turn off tracing but nothing else).

I'll file this under the list of things that we should fix, but aren't needed to land python functionalization though. So if no tests are passing, and I get the same trace out with C++ vs python functionalization, I'll leave it for now.

zou3519 replied:
It looks like this is because DispatchKey::Functionalize is a part of the CompositeImplicitAutograd keyset

bdhirsh replied:
sync'd offline - totally agreed, will change this (although in the near future when we turn on functionalization for pre_dispatch, we'll want a way to let functionalization not run these decomps)

bdhirsh commented on c10/core/impl/LocalDispatchKeySet.h:
```diff
@@ -122,6 +122,15 @@ struct C10_API ForceDispatchKeyGuard {
       : saved_keyset_(c10::impl::tls_local_dispatch_key_set()) {
     c10::impl::_force_tls_local_dispatch_key_set(key_set);
   }
+  ForceDispatchKeyGuard(
```
agreed, will change

bdhirsh replied:
oh I just realized - this is the existing name, I just added an overload to the constructor.

bdhirsh commented on torch/_subclasses/functional_tensor.py:
```diff
@@ -28,6 +28,28 @@ class FunctionalTensor(torch.Tensor):
+        #     } else {
+        #         return at::view_as_real(x);
+        #     }
+        extra_dispatch_keys = torch._C.DispatchKeySet(
+            torch._C.DispatchKey.Conjugate
+        ).add(torch._C.DispatchKey.Negative)
+        extra_dispatch_keys = extra_dispatch_keys & torch._C._dispatch_keys(elem)
```
Oh nice - I'll just reuse this (and bind it to python somewhere). Yes, the CPU/CUDA/XLA keys get propagated like you said.

bdhirsh commented on test/test_functionalization.py:
```diff
@@ -1542,6 +1542,35 @@ def forward(self, x_1):
+    def test_python_functionalization_conj(self):
+        def f(x):
+            y = x.conj()
+            return torch.view_as_real(y.resolve_conj())
+
+        x = torch.randn(4, dtype=torch.complex64)
+        out_ref = f(x)
+        out_test = dispatch_functionalize(f)(x)
+        self.assertEqual(out_ref, out_test)
```
yep will do

bdhirsh committed:
Update on "Make FunctionalTensor subclass to be more like functorch (interaction with ZeroTensor + Conjugate key)"

This is a potential fix for https://github.com/pytorch/pytorch/issues/108983. It fixes an ugly segfault that I got when running a functorch `jacfwd()` test with `torch.compile`, once I started using `FunctionalTensor` in AOTAutograd.

Notable changes:

(1) register a fallthrough for ZeroTensor to `prim::clone`, which is what we do for `aten::clone`. We need this now if we expect the `clone()` decomp that we use to work with ZeroTensor's.

(2) `maybe_add_maybe_sub` is a kernel used by ZeroTensor that does some meta-tensor compute to figure out the size of the ZeroTensor to return. This broke in my tests because we end up dispatching `aten::to()` directly to my (FunctionalTensor) subclass. This is not ideal, because `FunctionalTensor` uses `return_and_correct_aliasing()` to properly handling storage aliasing, and to do that it reads aliasing information off of the schema of each op it sees. `to()` is a "maybe aliasing" op, and so we'd like to decompose it before getting to `_torch_dispatch_`

(3) I removed `ZeroTensor` (and conj and neg) from the exclude set in the python fallback kernel. This didn't feel too unreasonable... but I'm wondering if anyone has an opinion about it. An alternative would be to not do this, and instead to have `FunctionalTensor` manually remove these keys from the exclude set upon redispatch (aka put the onus on the tensor subclass to fix up TLS, as described in https://github.com/pytorch/pytorch/issues/108983).




[ghstack-poisoned]

bdhirsh committed:
Update on "Make FunctionalTensor subclass to be more like functorch (interaction with ZeroTensor + Conjugate key)"

This is a potential fix for https://github.com/pytorch/pytorch/issues/108983. It fixes an ugly segfault that I got when running a functorch `jacfwd()` test with `torch.compile`, once I started using `FunctionalTensor` in AOTAutograd.

Notable changes:

(1) register a fallthrough for ZeroTensor to `prim::clone`, which is what we do for `aten::clone`. We need this now if we expect the `clone()` decomp that we use to work with ZeroTensor's.

(2) `maybe_add_maybe_sub` is a kernel used by ZeroTensor that does some meta-tensor compute to figure out the size of the ZeroTensor to return. This broke in my tests because we end up dispatching `aten::to()` directly to my (FunctionalTensor) subclass. This is not ideal, because `FunctionalTensor` uses `return_and_correct_aliasing()` to properly handling storage aliasing, and to do that it reads aliasing information off of the schema of each op it sees. `to()` is a "maybe aliasing" op, and so we'd like to decompose it before getting to `_torch_dispatch_`

(3) I removed `ZeroTensor` (and conj and neg) from the exclude set in the python fallback kernel. This didn't feel too unreasonable... but I'm wondering if anyone has an opinion about it. An alternative would be to not do this, and instead to have `FunctionalTensor` manually remove these keys from the exclude set upon redispatch (aka put the onus on the tensor subclass to fix up TLS, as described in https://github.com/pytorch/pytorch/issues/108983).




[ghstack-poisoned]

bdhirsh committed:
Update on "Make FunctionalTensor subclass to be more like functorch (interaction with ZeroTensor + Conjugate key)"

This is a potential fix for https://github.com/pytorch/pytorch/issues/108983. It fixes an ugly segfault that I got when running a functorch `jacfwd()` test with `torch.compile`, once I started using `FunctionalTensor` in AOTAutograd.

Notable changes:

(1) register a fallthrough for ZeroTensor to `prim::clone`, which is what we do for `aten::clone`. We need this now if we expect the `clone()` decomp that we use to work with ZeroTensor's.

(2) `maybe_add_maybe_sub` is a kernel used by ZeroTensor that does some meta-tensor compute to figure out the size of the ZeroTensor to return. This broke in my tests because we end up dispatching `aten::to()` directly to my (FunctionalTensor) subclass. This is not ideal, because `FunctionalTensor` uses `return_and_correct_aliasing()` to properly handling storage aliasing, and to do that it reads aliasing information off of the schema of each op it sees. `to()` is a "maybe aliasing" op, and so we'd like to decompose it before getting to `_torch_dispatch_`

(3) I removed `ZeroTensor` (and conj and neg) from the exclude set in the python fallback kernel. This didn't feel too unreasonable... but I'm wondering if anyone has an opinion about it. An alternative would be to not do this, and instead to have `FunctionalTensor` manually remove these keys from the exclude set upon redispatch (aka put the onus on the tensor subclass to fix up TLS, as described in https://github.com/pytorch/pytorch/issues/108983).




[ghstack-poisoned]

bdhirsh committed:
Update on "Make FunctionalTensor subclass to be more like functorch (interaction with ZeroTensor + Conjugate key)"

This is a potential fix for https://github.com/pytorch/pytorch/issues/108983. It fixes an ugly segfault that I got when running a functorch `jacfwd()` test with `torch.compile`, once I started using `FunctionalTensor` in AOTAutograd.

Notable changes:

(1) register a fallthrough for ZeroTensor to `prim::clone`, which is what we do for `aten::clone`. We need this now if we expect the `clone()` decomp that we use to work with ZeroTensor's.

(2) `maybe_add_maybe_sub` is a kernel used by ZeroTensor that does some meta-tensor compute to figure out the size of the ZeroTensor to return. This broke in my tests because we end up dispatching `aten::to()` directly to my (FunctionalTensor) subclass. This is not ideal, because `FunctionalTensor` uses `return_and_correct_aliasing()` to properly handling storage aliasing, and to do that it reads aliasing information off of the schema of each op it sees. `to()` is a "maybe aliasing" op, and so we'd like to decompose it before getting to `_torch_dispatch_`

(3) I removed `ZeroTensor` (and conj and neg) from the exclude set in the python fallback kernel. This didn't feel too unreasonable... but I'm wondering if anyone has an opinion about it. An alternative would be to not do this, and instead to have `FunctionalTensor` manually remove these keys from the exclude set upon redispatch (aka put the onus on the tensor subclass to fix up TLS, as described in https://github.com/pytorch/pytorch/issues/108983).




[ghstack-poisoned]

bdhirsh committed:
Update on "Make FunctionalTensor subclass to be more like functorch (interaction with ZeroTensor + Conjugate key)"

This is a potential fix for https://github.com/pytorch/pytorch/issues/108983. It fixes an ugly segfault that I got when running a functorch `jacfwd()` test with `torch.compile`, once I started using `FunctionalTensor` in AOTAutograd.

Notable changes:

(1) register a fallthrough for ZeroTensor to `prim::clone`, which is what we do for `aten::clone`. We need this now if we expect the `clone()` decomp that we use to work with ZeroTensor's.

(2) `maybe_add_maybe_sub` is a kernel used by ZeroTensor that does some meta-tensor compute to figure out the size of the ZeroTensor to return. This broke in my tests because we end up dispatching `aten::to()` directly to my (FunctionalTensor) subclass. This is not ideal, because `FunctionalTensor` uses `return_and_correct_aliasing()` to properly handling storage aliasing, and to do that it reads aliasing information off of the schema of each op it sees. `to()` is a "maybe aliasing" op, and so we'd like to decompose it before getting to `_torch_dispatch_`

(3) I removed `ZeroTensor` (and conj and neg) from the exclude set in the python fallback kernel. This didn't feel too unreasonable... but I'm wondering if anyone has an opinion about it. An alternative would be to not do this, and instead to have `FunctionalTensor` manually remove these keys from the exclude set upon redispatch (aka put the onus on the tensor subclass to fix up TLS, as described in https://github.com/pytorch/pytorch/issues/108983).




[ghstack-poisoned]

zou3519 commented on test/test_functionalization.py:
```diff
@@ -1542,6 +1543,87 @@ def forward(self, x_1):
+        out_test_cpp = _functionalize(f, reapply_views=True, crossref=False, skip_input_mutations=True)(x)
+        self.assertEqual(out_ref, out_test)
+        self.assertEqual(out_ref, out_test_cpp)
+        if not IS_WINDOWS:
```
What does windows have to do with it?

bdhirsh replied:
whoops I'll take it out - the test right above it had the same check, although looking at the PR history I realize now that this is only for `instance_norm` (strangely when we trace instance_norm, the nodes in the FX graph show up in a slightly different order on windows, breaking expect tests) https://github.com/pytorch/pytorch/pull/88697/files#diff-e302079b2ae45f6ada1a3cf98a49dfe5b518efbb51d2299de9d494246d4f74cbR1238

zou3519 commented on test/test_functionalization.py:
```diff
@@ -1542,6 +1543,87 @@ def forward(self, x_1):
+    zeros = torch.ops.aten.zeros.default([], dtype = torch.float32, layout = torch.strided, device = device(type='cpu'))
+    expand = torch.ops.aten.expand.default(zeros, [4]);  zeros = None
+    _to_copy_1 = torch.ops.aten._to_copy.default(expand, device = device(type='meta'), pin_memory = False);  expand = None
+    add = torch.ops.prims.add.default(_to_copy, _to_copy_1);  _to_copy = _to_copy_1 = None
```
Why do we get torch.ops.prims in the trace?

bdhirsh replied:
I had to double check, this is a good q. It looks like it's because ZeroTensor implements most binary ops by redispatching to the op with meta tensors ([code](https://github.com/pytorch/pytorch/blob/main/aten/src/ATen/native/BinaryOps.cpp#L1019)).

Since this is a direct redispatch, we go **straight** to the meta implementation of add/mul/sub, without hitting the python key first.

This invokes the meta implementation of `add()`, which we currently register from python, and calls the primtorch implementation ([code](https://github.com/pytorch/pytorch/blob/main/torch/_refs/__init__.py#L1080)).

I think in theory, we'd probably want any meta-compute that zero-tensor does to be const-propp'd and not bake the ops into the graph (future fix?).

zou3519 commented on test/test_functionalization.py:
```diff
@@ -1542,6 +1543,87 @@ def forward(self, x_1):
+        out_ref = f(x)
+        out_test = dispatch_functionalize(f)(x)
+        out_test_cpp = _functionalize(f, reapply_views=True, crossref=False, skip_input_mutations=True)(x)
+        # TODO: this is currently wrong, file an issue
+        # self.assertEqual(out_ref, out_test)
+        self.assertEqual(out_test_cpp, out_test_cpp)
```
Add a separate test for this and add a unittest.expectedFailure?

bdhirsh replied:
agreed, will separate out

bdhirsh replied:
nvm this test was buggy (I need to clone x first since otherwise the test mutates its input - so i'll get different results if I run it repeatedly)

zou3519 commented on test/test_functionalization.py:
```diff
@@ -1542,6 +1543,87 @@ def forward(self, x_1):
+        out_ref = f(x)
+        out_test = dispatch_functionalize(f)(x)
+        out_test_cpp = _functionalize(f, reapply_views=True, crossref=False, skip_input_mutations=True)(x)
+        # TODO: this is currently wrong, file an issue
+        # self.assertEqual(out_ref, out_test)
```
Which one is wrong? What does it return?

bdhirsh replied:
This is testing no functionalization vs. C++ functionalization.

So the C++ and python functionalization are both wrong (and give equally wrong results)

zou3519 commented on test/test_functionalization.py:
```diff
@@ -1448,6 +1448,7 @@ def forward(self, arg0_1, arg1_1, arg2_1):
     return view_5
     """)  # noqa: B950
 
+
```
You also probably want is_conj / is_neg unit tests. I'm convinced they work, but not bad to test them so it's easy for someone who breaks the code in the future to figure out.

zou3519 commented on torch/_subclasses/functional_tensor.py:
```diff
@@ -29,8 +29,36 @@ class FunctionalTensor(torch.Tensor):
     # this is an "infra" mode with lower dispatching precedence.
     _mode_key = torch._C._TorchDispatchModeKey.FUNCTIONAL
 
+    # Note: FunctionalTensor is not quite like the functorch wrapper subclasses,
+    # because today functionalization runs *below* the conj/neg/zero dispatch keys.
+    # We could consider changing this to make it more inline with functorch.
+    _extra_dispatch_keys = torch._C._additional_keys_to_prop_for_wrapper_tensors.add(
+        torch._C.DispatchKey.ZeroTensor
+    )
```
nit: The comment doesn't really explain what we're doing here.

bdhirsh replied:
Fair, I'll make this clearer

zou3519 commented on torch/_subclasses/functional_tensor.py:
```diff
@@ -29,8 +29,36 @@ class FunctionalTensor(torch.Tensor):
+        #     } else {
+        #         return at::view_as_real(x);
+        #     }
+        extra_dispatch_keys = (
+            FunctionalTensor._extra_dispatch_keys & torch._C._dispatch_keys(elem)
+        )
```
You're also adding ZeroTensor here. I'm not sure that's what we want (we don't do this in functorch). we probably don't want to add ZeroTensor here, but we do want to re-enable (Conj, Neg, ZeroTensor, Functionalize) later on

bdhirsh replied:
I actually wanted this, to mirror the fact that the C++ `FunctionalTensorWrapper` will also propagate the ZeroTensor key from the inner tensor to the wrapper ([code](https://github.com/pytorch/pytorch/blob/main/aten/src/ATen/FunctionalTensorWrapper.cpp#L36)).

Something does seem fishy about doing this, although it's enough to get tests passing / get parity with C++ functionalization.

zou3519 replied:
Fair enough

zou3519 commented on torch/_subclasses/functional_tensor.py:
```diff
@@ -205,11 +237,19 @@ def unwrap(x):
             torch._C.DispatchKey.Functionalize
         )
         assert is_excluded or not is_included
+        include_to_set = (
+            torch._C._dispatch_tls_local_include_set()
+            | torch._C.DispatchKeySet(torch._C.DispatchKey.Functionalize)
+        )
+        exclude_to_set = (
+            torch._C._dispatch_tls_local_exclude_set().remove(
+                torch._C.DispatchKey.Functionalize
+            )
+            - FunctionalTensor._extra_dispatch_keys
+        )
```
NB: your PR message is saying that we did this in PythonFallBackKernel.cpp, but we're actually manually doing it in the subclass :P

bdhirsh replied:
sigh totally forgot to update the PR description, will update

zou3519 reviewed (commented):
I think the goal of this PR is to put up a Python functionalization that behaves the same as C++ functionalization (correct me if I'm wrong, the other goal would be to "figure out what we want functionalization to do when faced with conj tensors and fix the problems there".

Here are some mismatches:
- DispatchKey::Functionalize gets CompositeImplicitAutograd decomps. I'm not sure we should be doing the funny business with maybe_add_maybe_sub; instead, Python FunctionalTensor should get CompositeImplicitAutograd decomps. There is no guarantee that maybe_add_maybe_sub is the only place where we call some Composite operators below Autograd.
- ~~is_conj on C++ FunctionalTensorWrapper [doesn't seem like it can return True](https://gist.github.com/zou3519/3b9d2ee8ae80f155d9f6fbdbfa02e8ff); I'm not convinced we propagate the conj/neg dispatch keys on C++ FunctionalTensorWrapper. It's OK if you want to propagate them in this PR (that is the morally correct thing to do), but if we want to reduce risk we shouldn't propagate them in this PR and instead do it in another one at the top of the stack as a "bug-fix"~~
- DispatchKey::Functionalize is allowed to re-dispatch to things above it. Tensor Subclasses are not allowed to do this (as per PythonFallbackKernel.cpp). In your PR, we've just manually re-enabled these keys in FunctionalTensor. Longer-term, we should either revisit the Tensor Subclass keys or figure out how to apply the functorch-like design, but both of those are open questions right now, so no action required

EDIT: I'm wrong on the second bullet point. C++ functionaltensorwrapper does propagate conj/neg bit. So only outstanding item is the first bullet.

bdhirsh committed:
Update on "Make FunctionalTensor subclass to be more like functorch (interaction with ZeroTensor + Conjugate key)"

This is a potential fix for https://github.com/pytorch/pytorch/issues/108983. It fixes an ugly segfault that I got when running a functorch `jacfwd()` test with `torch.compile`, once I started using `FunctionalTensor` in AOTAutograd.

Notable changes:

(1) register a fallthrough for ZeroTensor to `prim::clone`, which is what we do for `aten::clone`. We need this now if we expect the `clone()` decomp that we use to work with ZeroTensor's.

(2) `maybe_add_maybe_sub` is a kernel used by ZeroTensor that does some meta-tensor compute to figure out the size of the ZeroTensor to return. This broke in my tests because we end up dispatching `aten::to()` directly to my (FunctionalTensor) subclass. This is not ideal, because `FunctionalTensor` uses `return_and_correct_aliasing()` to properly handling storage aliasing, and to do that it reads aliasing information off of the schema of each op it sees. `to()` is a "maybe aliasing" op, and so we'd like to decompose it before getting to `_torch_dispatch_`

(3) I removed `ZeroTensor` (and conj and neg) from the exclude set in the python fallback kernel. This didn't feel too unreasonable... but I'm wondering if anyone has an opinion about it. An alternative would be to not do this, and instead to have `FunctionalTensor` manually remove these keys from the exclude set upon redispatch (aka put the onus on the tensor subclass to fix up TLS, as described in https://github.com/pytorch/pytorch/issues/108983).




[ghstack-poisoned]

bdhirsh commented:
Updated based on the feedback. Change are:

(1) Removed the custom change to the zero tensor kernels, and instead updated `FunctionalTensorMode` to use `.decompose()` to run CompositeImplicitAutograd decomps when necessary (similar to FakeTensor).

(2) removed the IS_WINDOWS checks, not needed

(3) added is_conj/neg tests

(4) updated some comments

bdhirsh committed:
Update on "Make FunctionalTensor subclass to be more like functorch (interaction with ZeroTensor + Conjugate key)"

This is a potential fix for https://github.com/pytorch/pytorch/issues/108983. It fixes an ugly segfault that I got when running a functorch `jacfwd()` test with `torch.compile`, once I started using `FunctionalTensor` in AOTAutograd.

Notable changes:

(1) register a fallthrough for ZeroTensor to `prim::clone`, which is what we do for `aten::clone`. We need this now if we expect the `clone()` decomp that we use to work with ZeroTensor's.

(2) `maybe_add_maybe_sub` is a kernel used by ZeroTensor that does some meta-tensor compute to figure out the size of the ZeroTensor to return. This broke in my tests because we end up dispatching `aten::to()` directly to my (FunctionalTensor) subclass. This is not ideal, because `FunctionalTensor` uses `return_and_correct_aliasing()` to properly handling storage aliasing, and to do that it reads aliasing information off of the schema of each op it sees. `to()` is a "maybe aliasing" op, and so we'd like to decompose it before getting to `_torch_dispatch_`

(3) I removed `ZeroTensor` (and conj and neg) from the exclude set in the python fallback kernel. This didn't feel too unreasonable... but I'm wondering if anyone has an opinion about it. An alternative would be to not do this, and instead to have `FunctionalTensor` manually remove these keys from the exclude set upon redispatch (aka put the onus on the tensor subclass to fix up TLS, as described in https://github.com/pytorch/pytorch/issues/108983).




[ghstack-poisoned]

bdhirsh committed:
Update on "Make FunctionalTensor subclass to be more like functorch (interaction with ZeroTensor + Conjugate key)"

I added some tests for Conj, Neg and ZeroTensor for both python and C++ functionalization. This also fixes a nasty segfult when running a functorch `jacfwd` test with `torch.compile`, once AOTAutograd is using `FunctionalTensor`.

Changes:

(1) I use Jeffrey's `make_wrapper_subclass(extra_dispatch_keys)` kwarg to plumb extra dispatch keys ontoto the wrapper, mirroring what C++ functionalization does (C++ functionalization will mirror all dispatch keys from the inner tensor to the wrapper, except for python and functorch keys).

(2) FunctionalTensorMode will decompose CompositeImplicitAutograd ops, since (for example) ZeroTensor kernels can send ops like `.to()` directly to the Python key. We'll need a way to toggle this later for pre-dispatch functionalization

(3) Bound `_ForceDispatchKeyGuard` and BatchedTensorImpl's dispatch keyset to python




[ghstack-poisoned]

bdhirsh committed:
Update on "Make FunctionalTensor subclass to be more like functorch (interaction with ZeroTensor + Conjugate key)"

I added some tests for Conj, Neg and ZeroTensor for both python and C++ functionalization. This also fixes a nasty segfult when running a functorch `jacfwd` test with `torch.compile`, once AOTAutograd is using `FunctionalTensor`.

Changes:

(1) I use Jeffrey's `make_wrapper_subclass(extra_dispatch_keys)` kwarg to plumb extra dispatch keys ontoto the wrapper, mirroring what C++ functionalization does (C++ functionalization will mirror all dispatch keys from the inner tensor to the wrapper, except for python and functorch keys).

(2) FunctionalTensorMode will decompose CompositeImplicitAutograd ops, since (for example) ZeroTensor kernels can send ops like `.to()` directly to the Python key. We'll need a way to toggle this later for pre-dispatch functionalization

(3) Bound `_ForceDispatchKeyGuard` and BatchedTensorImpl's dispatch keyset to python




[ghstack-poisoned]

bdhirsh commented:
@pytorchbot merge
